File: solutions/go/square-root/1/square_root.go
package squareroot

func SquareRoot(number int) (int, error) {
	lower := 1
	upper := 256

	for {
		mid := (upper + lower) / 2
		square := mid * mid

		if square < number {
			lower = mid
		} else if square > number {
			upper = mid
		} else {
			return mid, nil
		}

		if lower >= upper-1 {
			break
		}
	}

	return lower, nil
}

File: solutions/r/square-root/1/square-root.R
square_root <- function(number) {
  if (number < 0) {
    stop("square root is not defined for negative numbers")
  }

  if (number == 0 || number == 1) {
    return(number)
  }

  guess <- number

  while (guess * guess > number) {
    guess <- (guess + number %/% guess) %/% 2
  }

  guess
}